native/dlahqr.go

```diff
@@ -0,0 +1,421 @@
+				// & Tisseur (LAWN 122, 1997). It has better
+				// mathematical foundation and improves accuracy
+				// in some cases.
+				if math.Abs(h[k*ldh+k-1]) <= ulp*tst {
+					ab := math.Max(math.Abs(h[k*ldh+k-1]), math.Abs(h[(k-1)*ldh+k]))
+					ba := math.Min(math.Abs(h[k*ldh+k-1]), math.Abs(h[(k-1)*ldh+k]))
+					aa := math.Max(math.Abs(h[k*ldh+k-0]), math.Abs(h[(k-1)*ldh+k-1]-h[k*ldh+k]))
+					bb := math.Min(math.Abs(h[k*ldh+k-0]), math.Abs(h[(k-1)*ldh+k-1]-h[k*ldh+k]))
+					s := aa + ab
+					if ab/s*ba <= math.Max(smlnum, aa/s*bb*ulp) {
+						break
+					}
+				}
+			}
+			l = k
+			if l > ilo {
+				// H[l,l-1] is negligible.
+				h[l*ldh+l-1] = 0
+			}
+			if l >= i-1 {
+				// Break the loop because a submatrix of order 1
+				// or 2 has split off.
+				converged = true
+				break
+			}
+
+			// Now the active submatrix is in rows and columns l to
+			// i. If eigenvalues only are being computed, only the
+			// active submatrix need be transformed.
+			if !wantt {
+				i1 = l
+				i2 = i
+			}
+
+			const (
+				dat1 = 3.0
+				dat2 = -0.4375
+			)
+			var h11, h21, h12, h22 float64
+			switch its {
+			case 10: // Exceptional shift.
+				s := math.Abs(h[(l+1)*ldh+l]) + math.Abs(h[(l+2)*ldh+l+1])
+				h11 = dat1*s + h[l*ldh+l]
+				h12 = dat2 * s
+				h21 = s
+				h22 = h11
+			case 20: // Exceptional shift.
+				s := math.Abs(h[i*ldh+i-1]) + math.Abs(h[(i-1)*ldh+i-2])
+				h11 = dat1*s + h[i*ldh+i]
+				h12 = dat2 * s
+				h21 = s
+				h22 = h11
+			default: // Prepare to use Francis' double shift (i.e.,
+				// 2nd degree generalized Rayleigh quotient).
+				h11 = h[(i-1)*ldh+i-1]
+				h21 = h[(i-0)*ldh+i-1]
+				h12 = h[(i-1)*ldh+i]
+				h22 = h[(i-0)*ldh+i]
+			}
+			s := math.Abs(h11) + math.Abs(h12) + math.Abs(h21) + math.Abs(h22)
+			var (
+				rt1r, rt1i float64
+				rt2r, rt2i float64
+			)
+			if s != 0 {
+				h11 /= s
+				h21 /= s
+				h12 /= s
+				h22 /= s
+				tr := (h11 + h22) / 2
+				det := (h11-tr)*(h22-tr) - h12*h21
+				rtdisc := math.Sqrt(math.Abs(det))
+				if det >= 0 {
+					// Complex conjugate shifts.
+					rt1r = tr * s
+					rt2r = rt1r
+					rt1i = rtdisc * s
+					rt2i = -rt1i
+				} else {
+					// Real shifts (use only one of them).
+					rt1r = tr + rtdisc
+					rt2r = tr - rtdisc
+					if math.Abs(rt1r-h22) <= math.Abs(rt2r-h22) {
+						rt1r *= s
+						rt2r = rt1r
+					} else {
+						rt2r *= s
+						rt1r = rt2r
+					}
+					rt1i = 0
+					rt2i = 0
+				}
+			}
+
+			// Look for two consecutive small subdiagonal elements.
+			var m int
+			for m = i - 2; m >= l; m-- {
+				// Determine the effect of starting the
+				// double-shift QR iteration at row m, and see
+				// if this would make H[m,m-1] negligible. The
+				// following uses scaling to avoid overflows and
+				// most underflows.
+				h21s := h[(m+1)*ldh+m]
+				s := math.Abs(h[m*ldh+m]-rt2r) + math.Abs(rt2i) + math.Abs(h21s)
+				h21s /= s
+				v1 := h21s*h[m*ldh+m+1] + (h[m*ldh+m]-rt1r)*((h[m*ldh+m]-rt2r)/s) - rt2i/s*rt1i
```

Comment:
This was something I wanted to revisit before submitting but apparently I did not. Moved `var v [3]float64` before this loop and started using it here.